source/language/types.rst

```diff
@@ -373,3 +375,152 @@ variables whose values may only be known at run time.
+| **bit**      | :gbg:`Yes` | :gbg:`Yes` | :gbg:`Yes` | :rbg:`No`   | :gbg:`Yes` | :center:`-`| :rbg:`No`  | :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **duration** | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No*`  | :rbg:`No`  | :rbg:`No`  | :center:`-`| :rbg:`No`  |
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+| **qubit**    | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :rbg:`No`   | :rbg:`No`  | :rbg:`No`  | :rbg:`No`  | :center:`-`|
++--------------+------------+------------+------------+-------------+------------+------------+------------+------------+
+
+\*Note: ``duration`` values can be converted to ``float`` using the division operator. See :ref:`divideDuration`
+
+Casting from bool
+~~~~~~~~~~~~~~~~~
+
+``bool`` values cast from ``false`` to ``0.0`` and from ``true`` to ``1.0`` or 
+an equivalent representation. ``bool`` values can only be cast to ``bit[1]`` 
+(a single bit), so explicit slicing syntax must be given if the target ``bit``
+has more than 1 bit of precision.
+
+Casting from int/uint
+~~~~~~~~~~~~~~~~~~~~~
+
+``int[n]`` and ``uint[n]`` values cast to the standard types mimicking ``C``
+behavior. Casting to ``bit[m]`` is only allowed when ``m==n``. If the target 
+``bit`` has more or less precision, then explicit slicing syntax must be given.
+
+Casting from float
+~~~~~~~~~~~~~~~~~~
+
+``float[n]`` values cast to the standard types mimicking ``C`` behavior. 
+Casting a ``float`` value to an ``angle[m]`` is accomplished by first 
+performing a modulo 2π operation on the float value. The resulting value 
+is then converted to the nearest ``angle[m]`` possible. In the event of a 
+tie between two possible nearest values the result is the one with an even 
+least significant bit (*i.e* round to nearest, ties to even).
+
+Casting from angle
+~~~~~~~~~~~~~~~~~~
+
+``angle[n]`` values cast to ``bool`` using the convention ``val != 0.0``. 
+Casting to ``float[m]`` is accomplished by first converting the angle value 
+to the nearest float, with ties going to the value with the even least 
+significant bit, and then multiplying the resulting value by 2π as a 
+``float[m]`` value. Casting to ``bit[m]`` values is only allowed when 
+``n==m``, otherwise explicit slicing synatx must be provided. 
+
+When casting between angles of differing precisions (``n!=m``): if the target 
+has more significant bits, then the value is padded with ``m-n`` least 
+significant bits of ``0``; if the target has fewer significant bits, then 
+there are two acceptable behaviors that can be supported by compilers: 
+rounding and truncation. For rounding the value is rounded to the nearest 
+value, with ties going to the value with the even least significant bit.
+Trunction is likely to have more hardware support. This behavior can be
+controlled by the use of a ``#pragma``.
```

Comment:
Should we talk with the implementation guys and try to decide on a roughlystandard syntax for the `#pragma` before merging this, or is it better to do later?  (While unknown pragmas can just be ignored, it may be better to try and nail down some of the basic ones first, even if just to decide on standard namespacing conventions.)

Reply:
**N.B.** There is a detail that we fretted about, concerning _how_ to cast an ``angle`` to the nearest possible ``float``. There are two subtleties: the fact that this is discontinuous around the value 0 or 2π, and the issue of doing it quickly in practise given that ``angle`` types are fixed precision, in contrast to IEEE 754. We should figure out how/where to note these things.

(The first is not _really_ a big deal, but is something we should openly remark on, because it is an obvious pitfall for a programmer to be aware of. For the second, I'm now sure that there is a simple procedure for involving just a few machine instructions; it's a question of where we want to note that.)

Reply:
Added in https://github.com/Qiskit/openqasm/pull/268/commits/8b7e89933298f4f3ffe4903e0d57743584aa9a1f

Reply:
We left the question of `#pragma` til later and the grammar working group, so this doesn't need to be added here.